diff --git a/trajopt/test/benchmarks/CMakeLists.txt b/trajopt/test/benchmarks/CMakeLists.txt
index f8cbb30..5f68f80 100644
--- a/trajopt/test/benchmarks/CMakeLists.txt
+++ b/trajopt/test/benchmarks/CMakeLists.txt
@@ -10,7 +10,7 @@ macro(add_benchmark benchmark_name benchmark_file)
   target_link_libraries(${benchmark_name}
       ${PROJECT_NAME}
       tesseract::tesseract
-      osqp::osqpstatic
+      osqp::osqp
       benchmark::benchmark
       )
   target_include_directories(${benchmark_name} PRIVATE
diff --git a/trajopt_sco/CMakeLists.txt b/trajopt_sco/CMakeLists.txt
index 373b213..0db669f 100644
--- a/trajopt_sco/CMakeLists.txt
+++ b/trajopt_sco/CMakeLists.txt
@@ -30,9 +30,10 @@ set(SCO_SOURCE_FILES
     src/num_diff.cpp
 )
 
-if (NOT APPLE AND NOT WIN32)
-  set (HAVE_BPMPD TRUE)
-endif()
+# if (NOT APPLE AND NOT WIN32)
+#  set (HAVE_BPMPD TRUE)
+# endif()
+set (HAVE_BPMPD FALSE)
 
 if (HAVE_BPMPD)
   add_executable(bpmpd_caller src/bpmpd_caller.cpp)
@@ -82,7 +83,7 @@ if (HAVE_BPMPD)
   target_compile_definitions(${PROJECT_NAME} PRIVATE BPMPD_CALLER="${CMAKE_INSTALL_PREFIX}/bin/bpmpd_caller" HAVE_BPMPD=ON)
 endif()
 if (osqp_FOUND)
-  target_link_libraries(${PROJECT_NAME} PRIVATE osqp::osqpstatic)
+  target_link_libraries(${PROJECT_NAME} PRIVATE osqp::osqp)
   target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_OSQP=ON)
 endif()
 if (qpOASES_FOUND)
diff --git a/trajopt_sco/test/CMakeLists.txt b/trajopt_sco/test/CMakeLists.txt
index 7c5ed8c..abe1352 100644
--- a/trajopt_sco/test/CMakeLists.txt
+++ b/trajopt_sco/test/CMakeLists.txt
@@ -40,7 +40,7 @@ set(SCO_TEST_SOURCE
 add_executable(${PROJECT_NAME}-test ${SCO_TEST_SOURCE})
 target_link_libraries(${PROJECT_NAME}-test GTest::GTest GTest::Main ${PROJECT_NAME})
 if (osqp_FOUND)
-    target_link_libraries(${PROJECT_NAME}-test osqp::osqpstatic)
+    target_link_libraries(${PROJECT_NAME}-test osqp::osqp)
 endif()
 target_compile_options(${PROJECT_NAME}-test PRIVATE ${TRAJOPT_COMPILE_OPTIONS_PRIVATE} ${TRAJOPT_COMPILE_OPTIONS_PUBLIC})
 target_compile_definitions(${PROJECT_NAME}-test PRIVATE ${TRAJOPT_COMPILE_DEFINITIONS} TRAJOPT_IFOPT_DIR="${CMAKE_SOURCE_DIR}")
